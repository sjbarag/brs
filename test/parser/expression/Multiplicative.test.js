const BrsError = require("../../../lib/Error");
const { Lexeme, BrsTypes, Parser } = require("brs");
const { Float } = BrsTypes;

const { EOF } = require("../ParserTests");

describe("parser", () => {
    afterEach(() => BrsError.reset());

    describe("multiplicative expressions", () => {
        it("parses left-associative multiplication chains", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Float, text: "3.0", literal: new Float(3.0), line: 1 },
                { kind: Lexeme.Star, text: "*", line: 1 },
                { kind: Lexeme.Float, text: "5.0", literal: new Float(5.0), line: 1 },
                { kind: Lexeme.Star, text: "*", line: 1 },
                { kind: Lexeme.Float, text: "7.0", literal: new Float(7.0), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses left-associative division chains", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Float, text: "7.0", literal: new Float(7.0), line: 1 },
                { kind: Lexeme.Slash, text: "/", line: 1 },
                { kind: Lexeme.Float, text: "5.0", literal: new Float(5.0), line: 1 },
                { kind: Lexeme.Slash, text: "/", line: 1 },
                { kind: Lexeme.Float, text: "3.0", literal: new Float(3.0), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses left-associative modulo chains", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Float, text: "7.0", literal: new Float(7.0), line: 1 },
                { kind: Lexeme.Mod, text: "MOD", line: 1 },
                { kind: Lexeme.Float, text: "5.0", literal: new Float(5.0), line: 1 },
                { kind: Lexeme.Mod, text: "MOD", line: 1 },
                { kind: Lexeme.Float, text: "3.0", literal: new Float(3.0), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });

        it("parses left-associative integer-division chains", () => {
            let parsed = Parser.parse([
                { kind: Lexeme.Identifier, text: "_", line: 1 },
                { kind: Lexeme.Equal, text: "=", line: 1 },
                { kind: Lexeme.Float, text: "32.5", literal: new Float(32.5), line: 1 },
                { kind: Lexeme.Backslash, text: "\\", line: 1 },
                { kind: Lexeme.Float, text: "5.0", literal: new Float(5.0), line: 1 },
                { kind: Lexeme.Backslash, text: "\\", line: 1 },
                { kind: Lexeme.Float, text: "3.0", literal: new Float(3.0), line: 1 },
                EOF
            ]);

            expect(BrsError.found()).toBeFalsy();
            expect(parsed).toBeDefined();
            expect(parsed).not.toBeNull();
            expect(parsed).toMatchSnapshot();
        });
    });
});
