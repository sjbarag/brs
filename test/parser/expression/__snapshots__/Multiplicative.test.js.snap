// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parser multiplicative expressions parses left-associative division chains 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Binary {
        "left": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 7,
          },
        },
        "right": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 5,
          },
        },
        "token": Object {
          "isReserved": false,
          "kind": "Slash",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "/",
        },
        "type": "Binary",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Float {
          "kind": 6,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Slash",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "/",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser multiplicative expressions parses left-associative integer-division chains 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Binary {
        "left": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 32.5,
          },
        },
        "right": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 5,
          },
        },
        "token": Object {
          "isReserved": false,
          "kind": "Backslash",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "\\\\",
        },
        "type": "Binary",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Float {
          "kind": 6,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Backslash",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "\\\\",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser multiplicative expressions parses left-associative modulo chains 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Binary {
        "left": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 7,
          },
        },
        "right": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 5,
          },
        },
        "token": Object {
          "isReserved": false,
          "kind": "Mod",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "MOD",
        },
        "type": "Binary",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Float {
          "kind": 6,
          "value": 3,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Mod",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "MOD",
      },
      "type": "Binary",
    },
  },
]
`;

exports[`parser multiplicative expressions parses left-associative multiplication chains 1`] = `
Array [
  Assignment {
    "name": Object {
      "isReserved": false,
      "kind": "Identifier",
      "literal": undefined,
      "location": Object {
        "end": Object {
          "column": -9,
          "line": -9,
        },
        "start": Object {
          "column": -9,
          "line": -9,
        },
      },
      "text": "_",
    },
    "tokens": Object {
      "equals": Object {
        "isReserved": false,
        "kind": "Equal",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "=",
      },
    },
    "type": "Assignment",
    "value": Binary {
      "left": Binary {
        "left": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 3,
          },
        },
        "right": Literal {
          "_location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "type": "Literal",
          "value": Float {
            "kind": 6,
            "value": 5,
          },
        },
        "token": Object {
          "isReserved": false,
          "kind": "Star",
          "literal": undefined,
          "location": Object {
            "end": Object {
              "column": -9,
              "line": -9,
            },
            "start": Object {
              "column": -9,
              "line": -9,
            },
          },
          "text": "*",
        },
        "type": "Binary",
      },
      "right": Literal {
        "_location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "type": "Literal",
        "value": Float {
          "kind": 6,
          "value": 7,
        },
      },
      "token": Object {
        "isReserved": false,
        "kind": "Star",
        "literal": undefined,
        "location": Object {
          "end": Object {
            "column": -9,
            "line": -9,
          },
          "start": Object {
            "column": -9,
            "line": -9,
          },
        },
        "text": "*",
      },
      "type": "Binary",
    },
  },
]
`;
